### DISCUSSION TITLE
Remove fields from CLR

### DISCUSSION THREAD

**Discussion by SetTrend** [dotnet/csharplang#4364] [Language Ideas]
Automatic properties and fields are basically the same.

So, if they are practically the same, why not be consistent and finally remove the formal distinction between Fields and Properties? It doesn't make sense to formally distinguish between both if there is no difference in the end.

So, I propose to remove fields from the CLR and let the current field syntax be the new syntax for automated properties:

![Replace automatic properties with fields](https://user-images.githubusercontent.com/9283914/106143261-d2754980-6172-11eb-9c47-7c729b6811be.gif)

This would be a great relief. We finally could get rid of this tedious, mindless, repetitive writing of `{ get; set; }` all the time and just hit a simple `;` instead.

It really doesn't make sense to distinguish the indistinguishable.

**Comment by En3Tho** [comment]
Be happy. Don't use properties.

**Reply by SetTrend** [reply]
That's not possible. Many Microsoft frameworks, like Entity Framework, WPF or UWP ignore fields.

**Comment by HaloFour** [comment]
They are completely different from a metadata proposition.  The fact that the JIT can elide the accessor methods in these cases is a runtime implementation detail.

Simply put, this proposal breaks everything about the runtime and all existing languages in the .NET ecosystem and there's less-than-a-zero percent chance of it ever being considered.

**Reply by SetTrend** [reply]
As stated in the [off-topic discussion](https://github.com/dotnet/csharplang/discussions/4359#discussioncomment-317703), there is no change in regard to auto-properties (i.e. properties without any code, without any access modifier or modifier whatsoever). So, there's nothing to break. Except for Fields becoming obsolete.

**Reply by HaloFour** [reply]
Fields aren't and will never be obsolete.  They are completely different from auto-properties.  The fact that the JIT can elide the accessors doesn't change that at all.

**Reply by SetTrend** [reply]
But only in concept, not in barebone reality.

**Reply by HaloFour** [reply]
The only reality is the metadata of the assembly.  What any given runtime happens to do with that assembly is entirely implementation details.  There is nothing in the ECMA CLI specs that states that a runtime has to inline those accessors.  That is very specifically an implementation detail of the runtime you happen to be using.

**Reply by GalaxiaGuy** [reply]
> But only in concept, not in barebone reality.

This seems to be the wrong way round. Auto properties and fields are conceptually the same, but in "barebone reality" very different.

**Reply by SetTrend** [reply]
How can they be different? I'm reading fluffy statements without hard facts. Would you please elaborate.

ECMA CLI specs were written decades ago. People back than had an idea that provided a solution for issues back then. But shortly after coming up with this spec, reality - apparently - proved their obsolescence.

The change wouldn't break anything. Behind the scenes, it's already compiled this way. This is already happening today. While we speak.

What's - in hard facts - keeping simple auto properties like `accessor datatype name { get; set; }` from being written like `accessor datatype name;` and field spec simply merged into these simple auto properties?

**Comment by huoyaoyuan** [comment]
Properties are **synthesized** while fields are **primitives**.
Properties are **contract** while fields are **implementation**.
Auto property just gives a handful way to synthesize common pattern. It cannot replace implementation.

**Reply by SetTrend** [reply]
That statement purely conceptual. Why shouldn't it be able to replace implementation?

**Reply by huoyaoyuan** [reply]
C# remains the ability to do low level manipulation with the runtime.
Fields are somewhere in the memory. It can be addressed, barriered, interlocked, and so on. If those abilities are provided on auto properties, they will become unique enough to be a separated concept.

**Reply by SetTrend** [reply]
But that excactly what can be done with simple auto properties, too. Auto properties with modifiers would still be kept from being addressed, barriered, interlocked, and so on. It's just the simple auto properties that would allow for that.

A member definition is also a **contract**<sup>*</sup>. You can't just change a property definition from one into the other without potentially breaking code in other assemblies.
<hr/>

<sup>*</sup> (anything beyond `private`)

**Comment by 333fred** [comment]
Kotlin actually did this: it has no fields, only properties, and optimizes private properties into fields as an implementation detail. However, doing so in C# would be a massive breaking change, so this is pretty much dead in the water.

**Reply by SetTrend** [reply]
Thank you, Fred.

I very much see you point.

Yet, I don't understand ... C#, the compiler and anything else is so agile to many regards. And with such a simple thing, we're pretty much dancing around the golden calf here.

**Reply by HaloFour** [reply]
It's not a simple change.  You're asking the runtime and entire language ecosystem to reconfigure itself around an erroneous understanding on your part, and at the end of all of that effort the _best_ case scenario is that we end right back up where we are today.  But even that requires adding dozens of new language and runtime features to be able to support all of the things you can do with fields that you can't do with properties but would now be required.  Fields and properties are fundamentally different concepts.  What the runtime may (or may not) do internally has absolutely no bearing on the conversation.

In short, this is a negative gain proposition because it involves tons of work by the runtime teams, the language teams, and every single C# project on the face of the planet, and at the end of all of that effort you gain absolutely nothing.

**Reply by saint4eva** [reply]
Using attribute, analyser or source generator could do that for you.

**Reply by saint4eva** [reply]
> I very much see you point.
>  And with such a simple thing, we're pretty much dancing around the golden calf here.

Such a simple thing indeed. Lol

**Reply by SetTrend** [reply]
So, you suggest to create additional resources and effort, just to circumvent what's already right in place?

**Reply by SetTrend** [reply]
> You're asking the runtime and entire language ecosystem to reconfigure

There is nothing to reconfigure. It's just IntelliSense and the compiler to respond appropriately, nothing else.

> to support all of the things you can do with fields that you can't do with properties

No, you're not seing the point:

- Properties will remain properties.
- Fields will be removed (thereby removing a severe field of ambiguity from the language).
- Only a simple language construct like `{ get; set; }` will be handled like a field. Just like it happens today.
- Any other property construct, like `{ get { ... }; set { ... }}` will not be touched.

> and at the end of all of that effort you gain absolutely nothing.

Wrong: This feature will save millions of time wasted on blatandly unnecessary typing `{ get; set; }` instead of just `;`.
<br/>

It is the same grade of improvement to the language as the invention of auto-properties was, replacing the tedious
```c#
private datatype _name;

public Name { get => _name; set => _name = value; }
```
... with just ...
```c#
public datatype Name { get; set; }
```
<br/>

That invention back then just wasn't consequenly done. Full consequence would have been to replace

```c#
private datatype _name;

public Name { get => _name; set => _name = value; }
```
... with ...
```c#
public datatype Name;
```

**Reply by SetTrend** [reply]
Regarding "removing fields":

During a transient phase, "field-like auto properties" could still be available with [`GetFields()`](https://docs.microsoft.com/dotnet/api/system.type.getfields#System_Type_GetFields) in Reflection. [`GetProperties()`](https://docs.microsoft.com/dotnet/api/system.type.getproperties) would simply list "field-like auto properties", too. If necessary, an additional Boolean [`PropertyInfo`](https://docs.microsoft.com/dotnet/api/system.reflection.propertyinfo) property could give information about whether a property is a simple auto property (i.e. "field-like auto property"), if that information isn't already available from any other [`PropertyInfo`](https://docs.microsoft.com/dotnet/api/system.reflection.propertyinfo) property anyway.

**Reply by HaloFour** [reply]
> There is nothing to reconfigure.

> Fields will be removed

If you can't see how "fields will be removed" is a massive "reconfiguration" then I'd suggest that you don't understand how fields work in .NET or what you can do with them.  How do you propose to take a `ref` to a property, let alone a native pointer?

> This feature will save millions of time wasted on blatandly unnecessary typing

And waste billions of hours in having to "reconfigure" the runtime and dozens of existing .NET languages.

This proposal will never happen.

**Comment by canton7** [comment]
Just because I don't think anyone's made a comprehensive list of stuff that will break yet...

1. `ref` works with fields but not properties. If you force fields to turn into properties, all usages of `ref` on them will break.
2. You can take the address of a field using `&`: not so with properties.
3. Reflection. You've covered this a bit, saying that `GetFields` would now return properties. People do however do things like `type.GetProperties()[0]`, and this could break.
6. There's also fun around using reflection to call into an older assembly, which does have distinct fields vs properties. Would `GetFields()` do the right things there?
3. IL uses different opcodes to access properties vs fields, so this is a binary-breaking change. 
4. This would also break anyone doing runtime code generation. They would still need to tell the difference between a field and a property.
4. Marshalling / pinvoke typically use fields. I'm honestly not sure what would happen if you started throwing properties around here.

I'm sure there are others: those are just the ones off the top of my head. I hope you appreciate that any attempt here would take a huge amount of time: both by the development team, and also by C# developers all over the world. Are you sure you wouldn't rather put that time towards something more useful?

**Reply by jnm2** [reply]
> Marshalling / pinvoke typically use fields. I'm honestly not sure what would happen if you started throwing properties around here.

In my experience, struct properties are invisible to p/invoke except for any backing fields they declare.

**Reply by canton7** [reply]
It looks like you can't apply `FieldOffset` to properties though, so you'd need to work around that. `StructLayoutAttribute.Pack` also only applies to fields.

**Reply by jnm2** [reply]
Wouldn't field-targeted attributes work on the backing fields of the properties, and wouldn't packing work as normal?

**Reply by canton7** [reply]
> Wouldn't field-targeted attributes work on the backing fields of the properties

Only if you keep the concept of a "field" or "backing field". It's not clear how far OP wants to take this, but my reading of it was that they want to remove the entire concept of a field ("Remove fields from the CLR"). That presumably also means the removal of field-targeted attributes, hence my comment about that needing to be worked around.

**Reply by SetTrend** [reply]
Given the idea of _simple auto properties replacing fields_ (i.e. _field-like properties_) existed,

>     1. `ref` works with fields but not properties. If you force fields to turn into properties, all usages of `ref` on them will break.

You could `ref` a construct like this: `public int myField {get; set;}` (because it would actually _be_ a field), though you couldn't `ref` a construct like this: `public int myProp {get; set { ... }}`.

(I'm using the moniker "field" here, but in the suggested change, fields would no-more exist because there'd only be properties.)

>     2. You can take the address of a field using `&`: not so with properties.

You could take the address using `&` of a construct like this: `public int myField {get; set;}` because it doesn't have a code body, though you couldn't take the address of construct like this: `public int myProp {get; set { ... }}`.

>     3. Reflection. You've covered this a bit, saying that `GetFields` would now return properties. People do however do things like `type.GetProperties()[0]`, and this could break.

Would you mind elaborating why this could break? The semantics wouldn't change for either method.

>     4. There's also fun around using reflection to call into an older assembly, which does have distinct fields vs properties. Would `GetFields()` do the right things there?

Both, `GetFields()` and `GetProperties()` may return a larger result set, because fields and _field-like properties_ wouldn't be distinguishable anymore. But given that everyone clung to style guides, stating that fields should be `private` or `protected` at most, while properties should be `public`, the actual risc level is medium to low.

Eventually, everything related to fields, i.e. [AttributeTargets.Field](https://docs.microsoft.com/dotnet/api/system.attributetargets), `GetFields()` et al. should generate a "deprecated" compiler warning/error and corresponding IL opcodes should be removed.

>     6. This would also break anyone doing runtime code generation. They would still need to tell the difference between a field and a property.

Just to make sure: Would they really _need_ to tell the difference? Or is this thought speculative?

>     7. Marshalling / pinvoke typically use fields. I'm honestly not sure what would happen if you started throwing properties around here.

For _field-like properties_, nothing would change.

**Reply by canton7** [reply]
> You could ref a construct like this: public int myField {get; set;} (because it would actually be a field), though you couldn't ref a construct like this: public int myProp {get; set { ... }}.

Right, but how would the calling code know that a property is `{ get; set; }` vs `{ get { ... } set { ... } }`? If it defined in another assembly you can't see the implementation. You've also got to worry about interfaces, virtual, etc.

I think that misses the wider point here, though. The entire point of using auto-properties is that you can later change the implementation of get/set *without a breaking change*. That's (the main reason) why people use auto-props over just using fields. If, as you propose, changing `get` into `get { ... }` becomes a breaking change, then you've lost the main reason to use auto-props in the first place! You can just use fields instead of auto-props, and have all of the same advantages and disadvantages as your proposal.

> You could take the address using & of a construct like this: public int myField {get; set;} because it doesn't have a code body, though you couldn't take the address of construct like this: public int myProp {get; set { ... }}.

Ditto. Just use a fields instead of this proposal if you're OK with this.

> Would you mind elaborating why this could break? The semantics wouldn't change for either method.

If `GetProperties` used to return 1 property, and now it returns 2, then `GetProperties()[0]` could return the wrong property. That will break some people.

> Just to make sure: Would they really need to tell the difference? Or is this thought speculative?

Yep, you do. `ldfld`/`stfld` vs `call`/`callvirt`.

> For field-like properties, nothing would change.

It seems all we're doing here is re-defining an auto-property to have exactly the same semantics as a field. Why not just use a field instead of using an auto-property? You seem to get all of the same advantages and disadvantages of your proposal.

The difference is that the rest of us can keep using auto-properties as a way to hide whether get/set has any logic in it from the caller, so we don't break the caller if we decide to change this later.

In other words, wouldn't banning auto-properties, and making people use fields instead, have all of the same effects as your proposal?

**Reply by SetTrend** [reply]
Yes, good thought.

**Reply by SetTrend** [reply]
Though,

> interfaces, virtual

would leave the realm of _field-like properties_, just like they do with fields, thereby prohibiting `ref`, `&` etc. on these.

The main issue is, that, currently, it's required to create a union of public fields and public auto properties to capture them all. MS Entity Framework (Core), ASP.NET (Core), WPF and UWP refrain from doing so and just enumerate public properties, but not public fields, in their implementation.

**Reply by SetTrend** [reply]
> Yep, you do. `ldfld`/`stfld` vs `call`/`callvirt`.

Not sure what you're implying with this statement. Which of these is called on today's simple auto properties?

See this spin-off from another discussion thread:

https://github.com/dotnet/csharplang/discussions/4359#discussioncomment-312255

**Reply by 333fred** [reply]
> Not sure what you're implying with this statement. Which of these is called on today's simple auto properties?

Property getters and setters are _always_ methods. There is no difference, to a consumer, between an autoprop and a regular prop. There is literally no way to tell which the class used without manual reflection. Both publicly and privately, all properties are fetched used call/callvirt because that's the point of properties. They're methods that encapsulate the internal implementation details.

**Reply by canton7** [reply]
> Not sure what you're implying with this statement. Which of these is called on today's simple auto properties?

`ldfld`/`stfld` and variants like `ldflda` work on fields. However you get or set a property by calilng methods, using `call`/`callvirt`. The thread you've linked to is about what the JIT emits, and the fact that it can inline simple property accesses. That's irrelevant to code which is doing runtime emission of IL.

Ultimately, though:

1. In the beginning, there were fields.
2. However, you can't add extra logic when someone tries to get/set a public field: you have to write a getter/setter method (think Java)
3. You can't change from using a public field to using a getter/setter method without that being a breaking change
4. So, everyone uses public getter/setter methods wrapping a private field (again think Java)
5. Properties are introduced to represent this concept and reduce boilerplate
6. Auto-properties are introduced to reduce the boilerplate for simple cases further

As far as I can tell, you're basically proposing to return to step 2, but instead of "fields" and "properties" you have "field-like properties" and ... "property-like properties"? You can't move from one type to the other type without that being a breaking change (step 3) so people move to always using "property-like properties" (step 4). And around we go.

Except, on the way we:

1. Introduce some means for consumers to tell the difference between field-like properties and property-like proeprties
2. Re-invent bits of the runtime (we need new mechanisms for doing `ref` on a field-like property, etc)
3. Massive binary break. "New" code can no longer call into "old" code.
4. Re-educate millions of C# programmers. Rewrite all of the tutorials which talk about fields.
5. Rethink a load of stuff like field-targetted attributes, interop, etc
6. Revamp bits of how reflection works on fields/properties

That's a massive massive undertaking (a finger-in-the-air estimate is at least hundreds of millions of dollars worth of time), for what? To end up back where we were before C# was born. You don't even have any guarantees that WPF et al would choose to bind to field-like properties as well as property-like properties!

Surely it would be a lot easier to add a bit to the reflection APIs to let it consume fields and properties with a unified interface, and convince WPF etc to use these?

**Reply by HaloFour** [reply]
All a "property" is in .NET is a name, type and a list of one or more methods.  One of those methods might be in the "getter" slot and/or one of those methods might be in the "setter" slot.  When you write code to read/write to those properties the compiler looks up the appropriate method and emits a call to that method:

```cs
person.Name = "Bill";
// is actually
person.SetName("Bill");
```

The compiler _must_ always emit a method call here.  It has no direct access to the backing field.  It doesn't even know if there is a backing field.  And even if it did, it's quite possible that it won't have that backing field tomorrow, or that the method might have additional logic.

Where the inlining comes into play is at runtime.  The runtime has the IL within the assemblies of both the caller and callee.  The runtime can make the determination that the callee is a simple method that meets the threshold to be inlined.  The runtime doesn't even care that the method is a property accessor method, let alone that the property happened to be a C# auto-property.  All it cares about is that the method is short and simple.  So the runtime, with the additional information that it does know, can inline the method, and if that method _happened_ to be a simple field access then the result would be as if the caller was accessing the backing field directly.  But the runtime could also do that for simple accessors that accessed multiple backing fields, or did other simple logic on those backing fields, as long as the method remained short and simple enough to fall under that threshold.  The actual threshold is runtime dependent.  Some runtimes could be more aggressive than others about if and when to inline.  Some runtimes might make a second pass at attempting to inline if the method method was found at runtime to be on a "hot" path.

TL;DR  Property read/writes are _always_ method invocations as far as the compiler is concerned (or aware).  Even if the compiler is working with a non-reference assembly that contains the real IL for the method it doesn't know for sure that this is the version of the assembly that will be supplied at runtime.  Those accessor methods, like any methods, can be short and sweet or long and complex.  The compiler won't know which, but the runtime will, which is why inlining happens at runtime.

**Reply by SetTrend** [reply]
@canton7, @HaloFour 

Very good arguments. 👍

Yes, I cannot find counter-arguments against these. Perfect.

Checkmate 👏 🤝

**Comment by AartBluestoke** [comment]
I like the conceptual difference between fields and properties.
Fields are where the data is stored.
(public) Properties are a contract for data that is trivially handed out by the class (get) or held(set).

That auto properties don't have a specific named place to store the data and can be considered to be holding the data is nice design, because there are often properties where the storage is different from what is publicly exposed.

```
class fooV1{
public int? Bar{get;setl;}
}

class fooV2{
private string _bar
public int? {get=>int.tryParse(_bar);set=>{_bar=value.ToString();}}
}
```

**Comment by TahirAhmadov** [comment]
I thought I was the one who proposed massive breaking changes around here. Looks like at least one other member beat me at it handily. :)
